### DISCUSSION TITLE
how to persist store data with indexed db?

### DISCUSSION THREAD

**Discussion by kunaaal13** [pmndrs/zustand#1901] [Q&A] [answered]
I'm currently working with a vite pwa react typescript app, so i want my states to be persisted in indexed db, so that i can use them when there is no network and i can't fetch the data from the server and set the states. I am using tanstack query for data ferching. Has anyone tried/managed to make this work?

**Comment by dbritto-dev** [comment]
> I'm currently working with a vite pwa react typescript app, so i want my states to be persisted in indexed db, so that i can use them when there is no network and i can't fetch the data from the server and set the states. I am using tanstack query for data ferching. Has anyone tried/managed to make this work?

@kunaaal13 if you are using tanstack query you do not need Zustand to cache data on indexed db. I would customize tanstack query to use indexed db to cache data

**Reply by kunaaal13** [reply]
So i had this data flow, where i thought, ill be getting data from the query and will set the persisted states, and can use them even if there is no data. Let's say im making an ordering an item flow where once an order is created ill update it to my server, but if im offline i can cache them into indexed db, and once im online i can sync them. So any idea how can i do this with zustand?

BTW, what is the tanstack query solution you are suggesting?

**Reply by dbritto-dev** [reply]
> So i had this data flow, where i thought, ill be getting data from the query and will set the persisted states, and can use them even if there is no data. Let's say im making an ordering an item flow where once an order is created ill update it to my server, but if im offline i can cache them into indexed db, and once im online i can sync them. So any idea how can i do this with zustand?
> 
> BTW, what is the tanstack query solution you are suggesting?

@kunaaal13 you said that you are using tanstack query so tanstack query is your data layer. I would stick with tanstack query since supports all of that and customize it to store the data on indexed db

BTW, it would be different if you are using Zustand as your data layer however you are using tanstack as your data layer so do not need Zustand for that job

**Reply by kunaaal13** [reply]
sounds great, so how can i customise the tanstack query to store the data to indexed db?

**Reply by dbritto-dev** [reply, marked as answer]
> sounds great, so how can i customise the tanstack query to store the data to indexed db? 

@kunaaal13 here you go:

- https://tanstack.com/query/v4/docs/react/plugins/persistQueryClient#persistqueryclientprovider
- https://tanstack.com/query/v4/docs/react/plugins/persistQueryClient#building-a-persister

**Reply by kunaaal13** [reply]
this i've read but again i had a few questions, like how to make it work and will it fulfill my requirements?

**Reply by dbritto-dev** [reply]
@kunaaal13 I would like to help just ping me

**Reply by kunaaal13** [reply]
@dbritto-dev sure
